File: frontend/lib/api/student-ai.ts
import { apiClient } from "./client";

export interface SourceCitation {
  resource_name: string;
  resource_id: string;
  page_number: number | null;
  chunk_index: number;
  excerpt: string;
}

export interface StudentSupportRequest {
  question: string;
  source_surface?: "study_tutor" | "study_reader" | "assessment_inline";
  conversation_id?: string;
  conversation_history?: Array<{ role: string; content: string }>;
  selected_resource_id?: string;
  selected_resource_ids?: string[];
  teaching_workspace_id?: string;
  thinking_mode?: boolean;
  deep_search_mode?: boolean;
  attempt_id?: string;
  question_id?: string;
  assessment_id?: string;
  is_in_assessment?: boolean;
  current_page?: number;
  selected_text?: string;
}

export interface StudentSupportResponse {
  explanation: string;
  conversation_id?: string;
  citations: SourceCitation[];
  fallback_used: boolean;
  model?: string | null;
  provider?: string | null;
}

export interface RevisionGuideOutput {
  title?: string;
  summary: string;
  checklist: string[];
  readings: string[];
  markdown?: string;
}

export interface StudentChatHistoryItem {
  id: string;
  conversation_id?: string;
  question: string;
  answer: string;
  citations: SourceCitation[];
  created_at: string;
}

export interface StudentConversationSummary {
  conversation_id: string;
  preview: string;
  created_at: string;
  last_activity_at: string;
  turn_count: number;
}

export const studentAiApi = {
  /**
   * Request study support from the AI agent.
   */
  async getSupport(data: StudentSupportRequest): Promise<StudentSupportResponse> {
    return apiClient("/student/ai/support", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },

  /**
   * Load student AI conversations summary list for sidebar history.
   */
  async getConversations(limit: number = 50, offset: number = 0): Promise<StudentConversationSummary[]> {
    return apiClient(`/student/ai/conversations?limit=${limit}&offset=${offset}`);
  },

  /**
   * Load all turns for a specific conversation thread.
   */
  async getConversation(conversationId: string): Promise<StudentChatHistoryItem[]> {
    return apiClient(`/student/ai/conversations/${conversationId}`);
  },

  /**
   * Delete a student AI conversation thread.
   */
  async deleteConversation(conversationId: string): Promise<{ success: boolean; conversation_id: string }> {
    return apiClient(`/student/ai/conversations/${conversationId}`, {
      method: "DELETE",
    });
  },

  /**
   * Load recent student AI chat history for conversation persistence (legacy).
   */
  async getHistory(): Promise<StudentChatHistoryItem[]> {
    return apiClient("/student/ai/history");
  },

  /**
   * Generate structured revision guide note and downloadable markdown sheet.
   */
  async getRevisionGuide(
    paramsOrTopic: string | { topic: string; teachingWorkspaceId?: string; learningUnitId?: string },
    teachingWorkspaceId?: string
  ): Promise<RevisionGuideOutput> {
    const payload = typeof paramsOrTopic === "string"
      ? { topic: paramsOrTopic, teaching_workspace_id: teachingWorkspaceId }
      : {
          topic: paramsOrTopic.topic,
          teaching_workspace_id: paramsOrTopic.teachingWorkspaceId,
          learning_unit_id: paramsOrTopic.learningUnitId,
        };
    return apiClient("/student/ai/revision", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  },
};

File: backend/app/api/v1/routes/student_ai.py
from __future__ import annotations

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.dependencies.auth import require_student
import uuid
from app.schemas.student_ai import (
    StudentSupportRequest,
    StudentSupportResponse,
    StudentChatHistoryItem,
    StudentConversationSummary,
    RevisionGuideRequest,
    RevisionGuideOutput,
)
from app.services.student_ai_service import StudentAIService

router = APIRouter(prefix="/student/ai", tags=["Student AI"])


@router.post(
    "/support",
    status_code=status.HTTP_200_OK,
    response_model=StudentSupportResponse,
    summary="Ask the Student Support Agent",
)
async def student_support(
    body: StudentSupportRequest,
    current_user=Depends(require_student),
    db: AsyncSession = Depends(get_db),
) -> StudentSupportResponse:
    service = StudentAIService(db)
    return await service.support(body, current_user)


@router.get(
    "/conversations",
    status_code=status.HTTP_200_OK,
    response_model=list[StudentConversationSummary],
    summary="Get student AI conversations summary list",
)
async def get_conversations(
    limit: int = 50,
    offset: int = 0,
    current_user=Depends(require_student),
    db: AsyncSession = Depends(get_db),
) -> list[StudentConversationSummary]:
    service = StudentAIService(db)
    return await service.get_conversations(current_user.id, limit=limit, offset=offset)


@router.get(
    "/conversations/{conversation_id}",
    status_code=status.HTTP_200_OK,
    response_model=list[StudentChatHistoryItem],
    summary="Get all turns in a student AI conversation thread",
)
async def get_conversation(
    conversation_id: uuid.UUID,
    current_user=Depends(require_student),
    db: AsyncSession = Depends(get_db),
) -> list[StudentChatHistoryItem]:
    service = StudentAIService(db)
    return await service.get_conversation(current_user.id, conversation_id)


@router.delete(
    "/conversations/{conversation_id}",
    status_code=status.HTTP_200_OK,
    summary="Delete a student AI conversation thread",
)
async def delete_conversation(
    conversation_id: uuid.UUID,
    current_user=Depends(require_student),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    service = StudentAIService(db)
    success = await service.delete_conversation(current_user.id, conversation_id)
    return {"success": success, "conversation_id": conversation_id}


@router.get(
    "/history",
    status_code=status.HTTP_200_OK,
    response_model=list[StudentChatHistoryItem],
    summary="Get student AI chat history (legacy)",
)
async def get_chat_history(
    current_user=Depends(require_student),
    db: AsyncSession = Depends(get_db),
) -> list[StudentChatHistoryItem]:
    service = StudentAIService(db)
    return await service.get_chat_history(current_user.id)


@router.post(
    "/revision",
    status_code=status.HTTP_200_OK,
    response_model=RevisionGuideOutput,
    summary="Generate structured revision guide",
)
async def generate_revision_guide(
    body: RevisionGuideRequest,
    current_user=Depends(require_student),
    db: AsyncSession = Depends(get_db),
) -> RevisionGuideOutput:
    service = StudentAIService(db)
    return await service.generate_revision_guide(body, current_user.id)
